### PR TITLE
Add callback in `lbfgs`

### PR DESCRIPTION
**Pull request by tmigot**
#33

### CODE REVIEW

**Comment by geoffroyleconte** [comment]
Hi, in Krylov.jl the callback returns a boolean, is there a reason why you chose to modify the status here?

**Comment by tmigot** [comment]
Personally, I think both options are possible.

I am following what has bien done for R2 https://github.com/JuliaSmoothOptimizers/JSOSolvers.jl/pull/134

**Comment by amontoison** [comment]
Maybe we should use the notation `callback!` if it updates the `solver` structure.

File: src/lbfgs.jl [hunk 1]
```diff
@@ -27,6 +27,22 @@ The keyword arguments may include
 # Output
 The returned value is a `GenericExecutionStats`, see `SolverCore.jl`.
 
+# Callback
+The callback is called after each iteration.
```
Comment (dpo) [inline src/lbfgs.jl:31]:
```suggestion
The callback is called at each iteration.
```

File: src/lbfgs.jl [hunk 2]
```diff
@@ -172,29 +209,35 @@ function SolverCore.solve!(
     ∇f .= ∇ft
 
     ∇fNorm = nrm2(n, ∇f)
-    iter = iter + 1
 
+    set_solution!(stats, x)
```
Comment (dpo) [inline src/lbfgs.jl:213]: This is costly to do at each iteration. The current iterate is already in the solver object. We should only store it in `stats` at the end.

**Review by dpo — CHANGES_REQUESTED** [review]

**Comment by tmigot** [comment]
> 

It is true that the callback can modify the anything... `callback` is actually the name of the keyword and I find it more user-friendly to be called `callback` instead of `callback!` but I see no problem in changing.

File: src/lbfgs.jl (hunk 1 above)
Comment (tmigot) [reply src/lbfgs.jl]: Thanks for the feedback. I also applied this to R2.

File: src/lbfgs.jl (hunk 2 above)
Comment (tmigot) [reply src/lbfgs.jl]: Thanks for the feedback. I also applied this to R2.

**Review by dpo — APPROVED** [review]

**Pull request merged**